### ISSUE TITLE
Event-bus protocol versioning / scoping

### ISSUE THREAD

**Issue by vietj** [eclipse-vertx/vert.x#3406]
Vert.x 4 will ship with a new version of the event-bus wire protocol as well as cluster manager topology usage. We should have a way to isolate / detect invalid protocols.

An approach could be to have each event-bus node declare the protocol it handles when it publish its node in the cluster manager.

A solution leveraging an initial handshake at the wire level would be more difficult to implement because the initial version does not have such handshake. The cluster manager declarative solution would have the benefit to detect invalid nodes earlier too.

**Comment by tsegismont** [comment]
## Current situtation

The format of messages sent from the event bus sender (client) to the recipient (server) is as follows:

```
+ length (int) + protocol version (byte) + payload (many fields) +
```

The server validates the protocol version when transforming a `Buffer` to a `ClusteredMessage` object:

- if the client version is greater than server version, the server throws an `IllegalStateException` (reported to the server context `exceptionHandler`)
- the client is not notified and fails with timeout

The server throws an ISE as well when it misses the message codec used by the client.

## Proposal

For `master` branch and then backport to `3.9` branch:

- create a new `io.vertx.core.eventbus.ReplyFailure`: `INTERNAL_ERROR`
- document that `INTERNAL_ERROR` is a fatal issue and there is no point in retrying to send the message
- server replies with `INTERNAL_ERROR` when protocol version is invalid or the message codec is missing

And then for `master` branch only:

- bump protocol version to `2` (4.0)

## Results

When 3.9.2 and below nodes connect to 3.9.3 and above nodes:

- if protocol version matches (eg 3.9.2 to 3.9.3), nothing happens
- if protocol version does not match (e.g. 3.9.2 to 4.0)
  - server node replies with `INTERNAL_ERROR`
  - client node fails with ISE (`INTERNAL_ERROR` is unknown) on the sender's context

When 3.9.3 and above nodes connect to 3.9.2 and below nodes:
- if protocol version matches (3.9.3 to 3.9.2), nothing happens
- if protocol version does not match (e.g. 4.0 to 3.9.2)
  - server node fails with ISE (reported to the server's context `exceptionHandler`)
  - the sender timeouts

When 3.9.3 and above nodes connect to 3.9.3 and above nodes:
- if protocol version matches (3.9.3 to 3.9.4), nothing happens
- if protocol version does not match (e.g. 3.9.3 to 4.0)
  - server node replies with `INTERNAL_ERROR`
  - sender receives reply failure

**Comment by rogelio-o** [comment]
Hi @tsegismont ,

Following up the conversation in vertx-dev group started by Mungai about migrating services at different times: 

Migrating using two cluster is hard for the following reasons: 
- One of the advantages of using messaging communication is transparency. You send the message to an address but you don't know what service (or services) is listen to the address. To migrate a service using two cluster you need to migrate all the services that are listening to an address in which the migrated service is posting messages (dependencies). Identify these dependencies is hard due to transparency.
- Following up dependencies you can find out that all services should be migrated at once. 
- All or nothing strategy is not possible in environment with tens of microservices managed by different teams.

I think with this approach migrating to version 4 will be hard (thus adoption to this version will be slow). Obviously, a new major version involves breaking changes, but it's important to notice that this breaking change is in communication (what implies coordination between deployments, extra development to achieve zero downtime, dependencies,...).

**Proposal**

What about backward compatibility? Supporting old protocol in version 4 and deprecate it in version 4.1. With this approach old and new protocol services can live together and it'll make migration smooth.

A way to do it (sending) could be:

1. Store node protocol version in `NodeInfo.metadata`. If there is no version stored, we can assume it's the old protocol because it's an old version node (without this mechanism).
2. Return node protocol in `selectForSend` and `selectForPublish` or retrieve it using the selected `nodeId` and the `ClusterManager`.
3. Use returned protocol to write message.

Backward compatibility should also be implemented in messages consumption. This can be done using protocol version indicated by the body.

What do you think?

**Comment by tsegismont** [comment]
@rogelio-o hi, I will be afk for 2 weeks, sorry for this short answer sent before logging off.

I need more time to think about your proposal for sharing protocol version in node metadata. But this is only one aspect of the issue: the information that is now published by the cluster manager is not the same, and not stored in the same maps/caches.

Also keep in mind that cluster managers themselves can be incompatible between versions. For example, Vert.x 3 works only with Hazelcast 3, and Vert.x 4 with Hazelcast 4. And even if Vert.x 4 was compatible with both Hazelcast 3.x and 4.x, you still wouldn't be able to mix nodes running different versions. So at some point you have to create a new cluster anyway.

If it's not possible to migrate all services at the same time, then perhaps using an event bus bridge could be useful.
I talked about this with @vietj today and I'll look into it when I'm back. This is certainly something we will document in the migration guide.

**Comment by rogelio-o** [comment]
Thanks for your answer @tsegismont! Enjoy your days off. Anything can help, just tell me.

**Comment by tsegismont** [comment]
@rogelio-o I sent an email to vertx-dev group about this

**Comment by tsegismont** [comment]
Closed in #3583 and bddb726
